src/bgp/aspath.rs

```diff
@@ -0,0 +1,1513 @@
+            _ => Err(InvalidSegmentTypeError)
+        }
+    }
+}
+
+impl<Octs> From<Segment<Octs>> for Hop<Octs> {
+    fn from(seg: Segment<Octs>) -> Self {
+        Hop::Segment(seg)
+    }
+}
+
+//--- OctetsFrom
+
+impl<Source, Octs> OctetsFrom<Hop<Source>> for Hop<Octs>
+    where
+    Octs: OctetsFrom<Source>
+{
+    type Error = Octs::Error;
+
+    fn try_octets_from(source: Hop<Source>) -> Result<Self, Self::Error> {
+        match source {
+            Hop::Asn(asn) => Ok(Hop::Asn(asn)),
+            Hop::Segment(seg) => Ok(
+                Hop::Segment(Segment::try_octets_from(seg)?)
+            )
+        }
+    }
+}
+
+//--- Display
+
+impl<Octs: Octets> fmt::Display for Hop<Octs> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Hop::Asn(a) => fmt::Display::fmt(a, f),
+            Hop::Segment(s) => fmt::Display::fmt(s, f)
+        }
+    }
+}
+
+
+//------------ Asns ----------------------------------------------------------
+
+/// Iterates over ASNs in a [`Segment`].
+pub struct Asns<'a, Octs> {
+    parser: Parser<'a, Octs>,
+    four_byte_asns: bool,
+}
+impl<'a, Octs> Asns<'a, Octs> {
+    fn new(parser: Parser<'a, Octs>, four_byte_asns: bool) -> Self {
+        Asns { parser, four_byte_asns }
+    }
+
+    fn into_segment(
+        mut self, stype: SegmentType
+    ) -> Segment<Octs::Range<'a>>
+    where Octs: Octets {
+        Segment::new(
+            stype,
+            self.four_byte_asns,
+            self.parser.parse_octets(
+                self.parser.remaining()
+            ).expect("parsed before")
+        )
+    }
+}
+
+impl<'a, Octs: Octets> Iterator for Asns<'a, Octs> {
+    type Item = Asn;
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.parser.remaining() == 0 {
+            return None
+        }
+        let n = if self.four_byte_asns {
+            self.parser.parse_u32().expect("parsed before")
+        }
+        else {
+            u32::from(self.parser.parse_u16().expect("parsed before"))
+        };
+        Some(Asn::from(n))
+    }
+}
+
+
+//------------ Helper Functions ----------------------------------------------
+
+fn asn_size(four_byte_asns: bool) -> usize {
+    if four_byte_asns {
+        4
+    }
+    else {
+        2
+    }
+}
+
+//============ Error Types ===================================================
+
+//------------ InvalidSegmentTypeError ---------------------------------------
+
+/// Error returned from conversions methods.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+pub struct InvalidSegmentTypeError;
+
+impl fmt::Display for InvalidSegmentTypeError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("invalid segment type")
+    }
+}
+
+impl error::Error for InvalidSegmentTypeError { }
+
+
+//------------ ToPathError ---------------------------------------------------
+// XXX is there a more idiomatic way of introducing another error next to
```

Comment:
Yes. Because `AppendError` is required to be `Into<ShortBuf>`, I generally make an enum that has `ShortBuf` as a variant.

Reply:
Yeah, that is an issue. I haven’t figured out a way to change the trait bounds to make this work.